mypy/newsemanal/semanal.py

```diff
@@ -4035,18 +4045,23 @@ def name_not_defined(self, name: str, ctx: Context, namespace: Optional[str] = N
                 self.add_fixture_note(fullname, ctx)
 
     def name_already_defined(self, name: str, ctx: Context,
-                    original_ctx: Optional[Union[SymbolTableNode, SymbolNode]] = None) -> None:
+                             original_ctx: Optional[Union[SymbolTableNode, SymbolNode]] = None
+                             ) -> None:
         if isinstance(original_ctx, SymbolTableNode):
-            node = original_ctx.node
+            node = original_ctx.node  # type: Optional[SymbolNode]
         elif isinstance(original_ctx, SymbolNode):
             node = original_ctx
+        else:
+            node = None
 
         if isinstance(original_ctx, SymbolTableNode) and isinstance(original_ctx.node, MypyFile):
             # Since this is an import, original_ctx.node points to the module definition.
             # Therefore its line number is always 1, which is not useful for this
             # error message.
             extra_msg = ' (by an import)'
-        elif node and node.line != -1:
+        elif node and node.line != -1 and node.fullname().startswith(self.cur_mod_id):
```

Comment:
A `startswith` check is not reliable. For example, `foo` is a prefix of `foobar`, even though they are unrelated names.